presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java

```diff
@@ -132,11 +133,40 @@ public void testSet()
         assertNoStateChange(stateMachine, () -> stateMachine.set(State.DINNER));
     }
 
+    @Test
+    public void testTerminalState()
+            throws Exception
+    {
+        StateMachine<Optional<Throwable>> stateMachine = new StateMachine<>("test", executor, Optional.empty(), Optional::isPresent);
+        assertEquals(stateMachine.get(), Optional.empty());
+
+        Optional<Throwable> value = Optional.of(new Throwable("This is terminal state value"));
+        Optional<Throwable> value2 = Optional.of(new Throwable("This is other terminal state value"));
+
+        assertNoStateChange(stateMachine, () -> stateMachine.compareAndSet(value2, value));
+
+        assertStateChange(stateMachine,
+                () -> stateMachine.compareAndSet(Optional.empty(), value),
+                value);
+
+        // attempt transition from a final state
+        assertNoStateChange(stateMachine, () -> {
+            try {
+                stateMachine.compareAndSet(value, value2);
+                fail("expected IllegalStateException");
+            }
+            catch (IllegalStateException expected) {
+            }
```

Comment:
you could validate the expected was raised:

`AtomicInteger expectedOccurred = new AtomicInteger();`

```suggestion
                expectedOccurred.incrementAndGet();
            }
```

+ `assertEquals(expectedOccurred.get(), 1);`

the existing tests can be improved too, as a separate preparatory commit

Reply:
Will do in n a separate commit / PR